tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+#include "tensorflow/core/util/env_var.h"
+
+namespace xla {
+namespace gpu {
+
+namespace {
+
+absl::InlinedVector<HloInstruction*, 2> GetOutputsOfFusion(
+    const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return {root};
+  } else {
+    return root->operands();
+  }
+}
+
+// Return the number of outputs of the fused computation.
+size_t GetOutputSizeOfFusion(const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return 1;
+  } else {
+    return ShapeUtil::TupleElementCount(root->shape());
+  }
+}
+
+class HorizontalFusionImpl {
+ public:
+  HorizontalFusionImpl(HloComputation* computation)
+      : computation_(computation) {}
+
+  ~HorizontalFusionImpl() {}
+
+  StatusOr<bool> Run();
+
+ private:
+  Status Fuse(absl::Span<HloInstruction*> fused_fusion_insns);
+
+  Status CreateFusedComputation(
+      absl::Span<HloInstruction*> fused_fusion_insns,
+      std::unique_ptr<HloComputation>* uniq_computation,
+      std::vector<HloInstruction*>* bound_operands);
+
+  // FusionCandidates collects profitable candidates for a given consumer
+  // instruction. GetNextSpanOfFusions() can then be iteratively invoked to
+  // acquire the next set of fusion candidates based on some heuristics.
+  class FusionCandidates {
+   public:
+    FusionCandidates(HloInstruction* consumer) : fusion_insns_(), pos_(0) {
+      Initialize(consumer);
+    }
+
+    // Get a span of fusions to be fused.
+    absl::Span<HloInstruction*> GetNextSpanOfFusions();
+
+   private:
+    void Initialize(HloInstruction*);
+
+    std::vector<HloInstruction*> fusion_insns_;
+    // pos_ points to the start position of next span.
+    size_t pos_;
+  };
+
+  HloComputation* computation_;
+};  // HorizontalFusionImpl
+
+void HorizontalFusionImpl::FusionCandidates::Initialize(
+    HloInstruction* consumer) {
+  // First, find out all fusion instructions. We will filter out
+  // unsupported/non-profitable cases below.
+  absl::flat_hash_set<HloInstruction*> fusion_insns;
+  for (auto opnd : consumer->operands()) {
+    auto predecessor = opnd->LatestNonGteAncestor();
+    if (predecessor->opcode() == HloOpcode::kFusion) {
+      fusion_insns.insert(predecessor);
+    }
+  }
+
+
+  auto is_supported_fusion = [&](const HloInstruction& insn) -> bool {
+    // Support only kLoop fusion.
+    if (!insn.IsLoopFusion()) {
+      return false;
+    }
+
+    return true;
+  };
+
+  std::function<bool(const HloInstruction&)> is_consumer_the_only_user =
+      [&](const HloInstruction& insn) {
+        return absl::c_all_of(insn.users(), [&](const HloInstruction* user) {
+          if (user->opcode() == HloOpcode::kGetTupleElement) {
+            // Skip GTE.
+            return is_consumer_the_only_user(*user);
+          } else if (user == consumer) {
```

Comment:
The large number of lambdas defined inside this function body isn't ideal for readability. In particular, this lambda captures `consumer` which is defined ~20 lines above and is called ~60 lines below. Please consider to extract the functions into an unnamed namespace and pass all arguments explicitly.

Reply:
Done.